components/scream/src/physics/p3/p3_functions_f90.cpp

```diff
@@ -2271,5 +2317,87 @@ void ice_water_conservation_f(Real qitot_, Real qidep_, Real qinuc_, Real qiberg
 
 }
 
+void check_values_f(Real* qv, Real* temp, Int kstart, Int kend,
+                    Int timestepcount, bool force_abort, Int source_ind, Real* col_loc)
+{
+  using P3F        = Functions<Real, DefaultDevice>;
+  using Spack      = typename P3F::Spack;
+  using Smask      = typename P3F::Smask;
+  using uview_1d   = typename P3F::uview_1d<Spack>;
+  using view_1d    = typename P3F::view_1d<Spack>;
+  using sview_1d   = typename P3F::view_1d<Smask>;
+  using usview_1d  = typename P3F::view_1d<Smask>;
+  using KT         = typename P3F::KT;
+  using ExeSpace   = typename KT::ExeSpace;
+  using MemberType = typename P3F::MemberType;
+
+  scream_require_msg(kend > kstart,
+                    "ktop must be larger than kstart, kstart, kend " << kend << kstart);
+  // bool
+  bool trap{false};
+  bool badvalue_found{false};
+
+  // zero based
+  kstart -= 1;
+  kend -= 1;
+ 
+  Int nk = kend - kstart + 1;
+
+  sview_1d qv_out_bounds_d("qv_d", nk);
+  const auto qv_out_bounds_h = Kokkos::create_mirror_view(qv_out_bounds_d);
+
+  sview_1d t_out_bounds_d("tv_d", nk);
+  const auto t_out_bounds_h = Kokkos::create_mirror_view(t_out_bounds_d);
+
+  // Set up input views
+  Kokkos::Array<view_1d, 2> t_d;
+  pack::host_to_device({qv, temp}, nk, t_d);
+
+  view_1d qv_d{t_d[0]},
+          temp_d{t_d[1]};
+
+  auto policy = util::ExeSpaceUtils<ExeSpace>::get_default_team_policy(1, nk);
+  Kokkos::parallel_for(policy, KOKKOS_LAMBDA(const MemberType& team) {
+
+    view_1d qv_d{t_d[0]},
+            temp_d{t_d[1]};
+
+    P3F::check_values(qv_d, temp_d, kstart, kend, team, qv_out_bounds_d, t_out_bounds_d);
+
+  });
+
+  Kokkos::deep_copy(qv_out_bounds_h, qv_out_bounds_d);
+  Kokkos::deep_copy(t_out_bounds_h, t_out_bounds_d);
+
```

Comment:
Everything below this line should be in the main implementation, not the bridge.